lightning/src/util/persist.rs

```diff
@@ -243,10 +250,886 @@ where
+		// Discard the sentinel bytes if found.
+		if monitor_cursor
+			.get_ref()
+			.starts_with(MONITOR_UPDATING_PERSISTER_PREPEND_SENTINEL)
+		{
+			monitor_cursor.set_position(MONITOR_UPDATING_PERSISTER_PREPEND_SENTINEL.len() as u64);
+		}
+		match <(
+			BlockHash,
+			ChannelMonitor<<SP::Target as SignerProvider>::Signer>,
+		)>::read(
+			&mut monitor_cursor,
+			(&*self.entropy_source, &*self.signer_provider),
+		) {
+			Ok((blockhash, channel_monitor)) => {
+				if channel_monitor.get_funding_txo().0.txid != outpoint.txid
+					|| channel_monitor.get_funding_txo().0.index != outpoint.index
+				{
+					log_error!(
+						self.logger,
+						"ChannelMonitor {} was stored under the wrong key!",
+						monitor_name.as_str()
+					);
+					return Err(io::Error::new(
+						io::ErrorKind::InvalidData,
+						"ChannelMonitor was stored under the wrong key",
+					));
+				}
+				Ok((blockhash, channel_monitor))
+			}
+			Err(e) => {
+				log_error!(
+					self.logger,
+					"Failed to read ChannelMonitor {}, reason: {}",
+					monitor_name.as_str(),
+					e,
+				);
+				Err(io::Error::new(
+					io::ErrorKind::InvalidData,
+					"Failed to read ChannelMonitor",
+				))
+			}
+		}
+	}
+
+	/// Read a channel monitor update.
+	fn read_monitor_update(
+		&self,
+		monitor_name: &MonitorName,
+		update_name: &UpdateName,
+	) -> io::Result<ChannelMonitorUpdate> {
+		let sub_ns = monitor_name.as_str();
+		let key = update_name.as_str();
+		let update_bytes =
+			self.kv_store
+				.read(CHANNEL_MONITOR_UPDATE_PERSISTENCE_NAMESPACE, sub_ns, key)?;
+		ChannelMonitorUpdate::read(&mut io::Cursor::new(update_bytes)).map_err(|e| {
+			log_error!(
+				self.logger,
+				"Failed to read ChannelMonitorUpdate {}/{}, reason: {}",
+				monitor_name.as_str(),
+				update_name.as_str(),
+				e,
+			);
+			io::Error::new(
+				io::ErrorKind::InvalidData,
+				"Failed to read ChannelMonitorUpdate",
+			)
+		})
+	}
+
+	/// Lazily delete updates in the given range for the given monitor.
+	fn lazy_delete_updates(
+		&self,
+		monitor_name: &MonitorName,
+		first_update_id: u64,
+		last_update_id: u64,
+	) -> io::Result<()> {
+		for update_id in first_update_id..=last_update_id {
+			let sub_ns = monitor_name.as_str();
+			let update_name = UpdateName::from(update_id);
+			debug_assert_eq!(
+				self.read_monitor_update(monitor_name, &update_name)?
+					.update_id,
+				update_name.0
+			);
+			self.kv_store.remove(
+				CHANNEL_MONITOR_UPDATE_PERSISTENCE_NAMESPACE,
+				sub_ns,
+				update_name.as_str(),
+				true,
+			)?;
+		}
+		Ok(())
+	}
+
+	/// Cleans up stale updates for all monitors, starting at a given search depth.
+	///
+	/// Usually, the value given for `maximum_pending_updates` in [`MonitorUpdatingPersister::new`]
+	/// is the best `search_depth` to use for routine cleaning, such as at startup, as stale updates
+	/// to that depth can be expected if there was a panic and crash.
+	///
+	/// "Deep cleaning" may be a useful, periodic maintenance task. Provide
```

Comment:
Projecting the routineness of my own crashes into the docs :) Yes, I think that is true, it would only come up if lazy deletes died in an internal queue or something like that. "Periodic" is probably not right for everyone.

Reply:
Revised / hopefully fixed.